### DISCUSSION TITLE
sitemap.xml cached somehow?

### DISCUSSION THREAD

**Discussion by darrylmorley** [vercel/next.js#56708] [Help]
### Summary

I'm trying to update my sitemap.xml, but it appears to be cached somehow as changes aren't reflected on my live site when updating with a new version.

I use an SEO tool generated sitemap.xml file, which is placed in the app directory. Pushed to Github and then the site is rebuilt on my Caprover server.

I've tested it in various browsers with the same result. Where could this cache issue be?

### Additional information

```js
next 13.4.19
caprover server - docker / ngninx
```

**Comment by neelratanguria** [comment]
Same problem, no help. I am using nextjs and vercel for deployment.

**Comment by fellmann** [comment]
Maybe a result of https://github.com/vercel/next.js/discussions/60998 ?

**Reply by neelratanguria** [reply]
I am using Nextjs 14.

I fixed the caching problem for API request using `export const dynamic = 'force-dynamic'`

Here is an example snippet
```

export async function GET(req, res) {
    try {
        // Make a GET request to the WorldTimeAPI
        const worldTimeResponse = await fetch('http://worldtimeapi.org/api/timezone/Asia/Kolkata', {
            cache:"no-cache"
        });
        
        if (!worldTimeResponse.ok) {
          throw new Error(`Failed to fetch data from WorldTimeAPI. Status: ${worldTimeResponse.status}`);
        }
    
        // Get the JSON response from the API
        const worldTimeData = await worldTimeResponse.json();
    
        // Convert the response data to a string (for demonstration purposes)
        const responseDataString = JSON.stringify(worldTimeData);
    
        // Return the response
        return new Response(responseDataString, { headers: { "Content-Type": "application/json" } });
      } catch (error) {
        // Handle errors, e.g., log the error and return an error response
        console.error('Error:', error.message);
        return new Response('Error fetching data from WorldTimeAPI', { status: 500 });
      }
}

export const dynamic = 'force-dynamic'

```

**Comment by jrhizor** [comment]
It looks like it only updates on deploy without `force-dynamic` for some reason. It really doesn't seem like that should be the case for `sitemap.ts` since there doesn't seem to be any other reason to have a programmatic sitemap?

**Comment by NabilNYMansour** [comment]
Just use [https://nextjs.org/docs/app/api-reference/functions/unstable_noStore](https://nextjs.org/docs/app/api-reference/functions/unstable_noStore)

For example:
```ts
import { myFetch } from '@/data';
import { unstable_noStore as noStore } from 'next/cache';

export default async function sitemap(): Promise<MetadataRoute.Sitemap> {
  noStore();
  const data = await myFetch();

  return ...
}
```

That worked fine for me. Though keep in mind that there is a limit of 50K links in a sitemap for Google. You will need to use [https://nextjs.org/docs/app/api-reference/functions/generate-sitemaps](https://nextjs.org/docs/app/api-reference/functions/generate-sitemaps) if you have more than that.

**Reply by fallenleavesguy** [reply]
Works well. awesome

**Comment by fellmann** [comment]
These solutions fix the issue, but not the underlying problem. Next.js framework should set caching headers correctly. It should not be necessary to fix it individually.

**Reply by NabilNYMansour** [reply]
My take is that Next.js should not cache until that is declared explicitly. So like instead of having a `noStore` function, we would have a `cachePage` function that needs to be declared on the top of the component. Or maybe something like `"use cache";` on the top of the file.

**Comment by HelaGone** [comment]
In my case our sitemaps are generated by the CMS and stored in a S3, but for some reason they got cached as well since they pass through the frontend domain and Idk how to prevent this to happen. Please help! 😢

**Comment by mboynes** [comment]
According to [the sitemap documentation](https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap):

> Good to know: sitemap.js is a special Route Handler that is cached by default unless it uses a [Dynamic API](https://nextjs.org/docs/app/building-your-application/caching#dynamic-apis) or [dynamic config](https://nextjs.org/docs/app/building-your-application/caching#segment-config-options) option.

**The answer to the larger question is that the sitemap route is cached (indefinitely) by next's Full Route Cache.** The Full Route Cache is cleared on new deployments. However, it should be noted that the Data Cache persists across deployments. Therefore, (as I understand it) it is theoretically possible that a Full Route Cache is regenerated after a deployment _using data in the Data Cache_.

In my use case, I didn't want to disable the Full Route Cache (or the Data Cache, for that matter); I want _some_ caching on the sitemap, just not _indefinite_ caching, since I'm pulling content dynamically from a CMS. I opted to go with using [the `revalidate` config option](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#revalidate) for the page, setting that to 5 minutes. Here's an example of that:

```ts
import { MetadataRoute } from 'next';

export const revalidate = 300;

export default async function sitemap(): Promise<MetadataRoute.Sitemap> {
  // ...
}
```

**Reply by joaottc** [reply]
unfortunately this doesn't seem to work, or at least it doesn't for netlify. Even with revalidate explicitly defined the sitemap continues to be cached

**Comment by nettum** [comment]
Yea, I'm having the same issue with cache components. 
Functions called from `sitemap.ts` uses `"use cache"` with mostly a short ttl.
There is no reason the sitemap should revalidate that often (which it does not do anyway) so I added the following to `sitemap.ts`:

```
export default async function sitemap(): Promise<MetadataRoute.Sitemap> {
  "use cache";
  cacheLife("hours");
 
 ....

}
```
This did of course not work (well no error, but you know - you find out eventually) as this is a special route that nextjs decides to cache forever.

So I did try to do the easy fix with `export const revalidate = 3600;`, but nah:

```
Error: Route segment config "revalidate" is not compatible with `nextConfig.cacheComponents`. Please remove it.
```

runtime APIs it is, I guess :man_shrugging:

**Comment by usualdesigner** [comment]
Since you ruled out the browser, I'd look at the Docker build itself before the app. If that sitemap.xml is a static file sitting in `public/` (not a generated route), it's easy for a `COPY public ./public` layer to get served from Docker's build cache if nothing above it in the Dockerfile changed — the file looks 'rebuilt' but the layer never actually re-copied. Try `docker build --no-cache` once and see if it fixes itself; if it does, the real fix is reordering your Dockerfile so that layer busts correctly. Second thing I'd check is whether CapRover's nginx has `proxy_cache` or `expires`/`Cache-Control` set globally for static assets — that'll happily serve a year-old sitemap even after a clean deploy.
